### PR TITLE
BE-953 add main services

### PR DESCRIPTION
**Pull request by DashaSyvyuk**

### CODE REVIEW

File: src/Entity/ApiKeyAccess.php [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+class ApiKeyAccess
+{
+    use DateStorageTrait;
+
+    const METHODS = [
+        'ANY'    => 'ANY',
+        'GET'    => 'GET',
+        'POST'   => 'POST',
+        'PUT'    => 'PUT',
+        'DELETE' => 'DELETE',
```
Comment (tomrutherford-livestyled) [inline src/Entity/ApiKeyAccess.php]: This array needs to include PATCH

File: src/EventBus/EventBridgeClient.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        try {
+            $result = $this->client->putEvents([
+                'Entries' => [
+                    [
+                        'Detail' => $detail,
+                        'DetailType' => $name,
+                        'EventBusName' => $this->getEventBusName(),
+                        'Resources' => ['apps/' . $app],
+                        'Source' => 'apiv3', // TODO payment
```
Comment (tomrutherford-livestyled) [inline src/EventBus/EventBridgeClient.php]: Source needs to change, please use `APP_NAME` from the `.env` file. This should be passed into the constructor like `env`

File: src/EventListener/EntityPreWriteSubscriber.php [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+use ApiPlatform\Core\EventListener\EventPriorities;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpKernel\Event\ViewEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+use Symfony\Component\PropertyAccess\PropertyAccessor;
+
+final class EntityPreWriteSubscriber implements EventSubscriberInterface
+{
+    // TODO remove
+    const APP_ID = 27;
```
Comment (tomrutherford-livestyled) [inline src/EventListener/EntityPreWriteSubscriber.php]: Why do we need these?

File: src/EventListener/EntityPreWriteSubscriber.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    public function attachApp(ViewEvent $event)
+    {
+        $entity = $event->getControllerResult();
+        $propertyAccessor = new PropertyAccessor();
+
+        if (!is_object($entity) || !$propertyAccessor->isWritable($entity, 'app')) {
+            return;
+        }
+
+        $propertyAccessor->setValue($entity, 'app', self::APP_ID);
```
Comment (tomrutherford-livestyled) [inline src/EventListener/EntityPreWriteSubscriber.php]: Why are we always setting app 27?

File: src/EventListener/EntityPreWriteSubscriber.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    public function attachTenant(ViewEvent $event)
+    {
+        $entity = $event->getControllerResult();
+        $propertyAccessor = new PropertyAccessor();
+
+        if (!$propertyAccessor->isWritable($entity, 'tenant')) {
+            return;
+        }
+
+        $propertyAccessor->setValue($entity, 'tenant', self::TENANT_ID);
```
Comment (tomrutherford-livestyled) [inline src/EventListener/EntityPreWriteSubscriber.php]: same here

File: src/Routing/IriConverter.php [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+     * @param string $resourceClass
+     * @return boolean
+     */
+    private function isMicroserviceAware(string $resourceClass): bool
+    {
+        try {
+            $class = new \ReflectionClass($resourceClass);
+            if ($this->reader->getClassAnnotation(
+                $class,
+                'LS\ApiBundle\Annotation\MicroserviceAware'
```
Comment (tomrutherford-livestyled) [inline src/Routing/IriConverter.php]: I think this namespace is now incorrect

File: src/Security/Configurator/Configurator.php [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    public function onKernelRequest(RequestEvent $event)
+    {
+        /** @var AppFilter $filter */
+        $filter = $this->em->getFilters()->enable('app_filter');
+
+        try {
+            $path = $event->getRequest()->getPathInfo();
+
+            if (strpos(trim($path, '/'), 'v4') === 0) {
+                $filter->setParameter('currentApp', self::CURRENT_APP);
```
Comment (tomrutherford-livestyled) [inline src/Security/Configurator/Configurator.php]: Why are we always setting 27?

File: src/Security/Extension/AppFilterExtension.php [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+    private function isApiRequest(array $context): bool
+    {
+        return (bool) preg_match('/\/v\d+\//', $context['request_uri'] ?? null);
+    }
+
+    private function enableAppFilter(): void
+    {
+        $filter = $this->em->getFilters()->enable('app_filter');
+        $filter->setParameter('currentApp', self::DEFAULT_APP);
```
Comment (tomrutherford-livestyled) [inline src/Security/Extension/AppFilterExtension.php]: same here

File: src/Validator/Translations.php [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+ */
+class Translations extends Constraint
+{
+    public string $message = '';
+
+    public function getTargets()
+    {
+        return self::PROPERTY_CONSTRAINT;
+    }
+}
```
Comment (tomrutherford-livestyled) [inline src/Validator/Translations.php]: need new line

File: composer.json [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+    "name": "realifetech/api-key-authenticator",
+    "description": "RealifeTech Api Key Authenticator",
+    "type": "library",
+    "license": "MIT",
+    "require": {
+        "php": ">=7.4",
+        "symfony/symfony": "^5.2",
+        "doctrine/orm": "^2.8",
+        "nesbot/carbon": "^2.46",
+        "api-platform/api-pack": "^1.3",
```
Comment (tomrutherford-livestyled) [inline composer.json:11]: Do we need Api Platform?

File: composer.json [hunk 11]
```diff
@@ -0,0 +1,19 @@
+{
+    "name": "realifetech/api-key-authenticator",
+    "description": "RealifeTech Api Key Authenticator",
+    "type": "library",
+    "license": "MIT",
+    "require": {
+        "php": ">=7.4",
+        "symfony/symfony": "^5.2",
```
Comment (tomrutherford-livestyled) [inline composer.json]: Symfony:4.4

File: src/Entity/ApiKey.php [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+     * @ORM\Id
+     * @ORM\GeneratedValue(strategy="AUTO")
+     * @ORM\Column(type="integer", options={"unsigned"=true})
+     */
+    private int $id;
+
+    /**
+     * @ORM\Column(type="integer", name="app_id")
+     */
+    private int $app;
```
Comment (tomrutherford-livestyled) [inline src/Entity/ApiKey.php]: Should be $tenant

**Review by tomrutherford-livestyled — CHANGES_REQUESTED** [review]

File: composer.json (hunk 10 above)
Comment (DashaSyvyuk) [reply composer.json:11]: we use it for IRI

File: src/EventBus/EventBridgeClient.php [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        try {
+            $result = $this->client->putEvents([
+                'Entries' => [
+                    [
+                        'Detail' => $detail,
+                        'DetailType' => $name,
+                        'EventBusName' => $this->getEventBusName(),
+                        'Resources' => ['apps/' . $tenant],
+                        'Source' => $this->env,
```
Comment (tomrutherford-livestyled) [inline src/EventBus/EventBridgeClient.php]: This should be the application name. We need another parameter in the constructor named `$source` and it should be set from the Symfont env variable `APP_NAME`

File: src/Provider/TenantProvider.php [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+    ) {
+        $this->apiKeyRepository = $apiKeyRepository;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function loadUserByUsername($username)
+    {
+        //pass
```
Comment (tomrutherford-livestyled) [inline src/Provider/TenantProvider.php]: Maybe we should throw an exception here

File: src/Provider/TenantProvider.php [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function refreshUser(UserInterface $user)
```
Comment (tomrutherford-livestyled) [inline src/Provider/TenantProvider.php:59]: and here

File: src/Provider/TenantProvider.php [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+    {
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function supportsClass($class)
```
Comment (tomrutherford-livestyled) [inline src/Provider/TenantProvider.php:72]: and here

File: src/Security/Filter/TenantFilter.php [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+        $query = '';
+
+        if (empty($this->reader)) {
+            return '';
+        }
+
+        if ($this->hasParameter('tenant')) {
+            $fieldName = 'tenant';
+        } else {
+            $fieldName = 'app_id';
```
Comment (tomrutherford-livestyled) [inline src/Security/Filter/TenantFilter.php]: should probably add a check for this field as well instead of assuming it's always there if `tenant` isnt

**Review by tomrutherford-livestyled — CHANGES_REQUESTED** [review]

**Review by tomrutherford-livestyled — APPROVED** [review]

**Pull request merged**